File: crates/flighthq-scene-gl/src/matcap_gl_mesh_material_renderer.rs
//! The built-in matcap forward renderer.
//!
//! Ports `@flighthq/scene-gl` `matcapGlMeshMaterialRenderer.ts`.

use flighthq_materials::unlit_materials::matcap_material_kind;
use flighthq_render_gl::GlRenderState;
use flighthq_types::camera::Camera;
use flighthq_types::scene_render::{SceneLightBlock, SceneRenderProxy};

use crate::gl_mesh_material_registry::{
    GlMeshMaterialRenderer, MeshMaterial, register_gl_mesh_material_renderer,
};
use crate::gl_scene_runtime::{GlMeshUpload, GlSceneRuntime};

/// The built-in matcap renderer.
pub struct MatcapGlMeshMaterialRenderer;

impl GlMeshMaterialRenderer for MatcapGlMeshMaterialRenderer {
    fn bind(
        &self,
        _state: &mut GlRenderState,
        _scene: &mut GlSceneRuntime,
        _material: Option<&dyn MeshMaterial>,
        _lights: &SceneLightBlock,
        _camera: &Camera,
    ) {
        // Stub: requires the family prelude program + material uniform upload.
    }

    fn draw(
        &self,
        _state: &mut GlRenderState,
        _scene: &mut GlSceneRuntime,
        _proxy: &SceneRenderProxy,
        _upload: &GlMeshUpload,
    ) {
        // Stub: requires draw_gl_mesh_subset.
    }
}

/// Registers the built-in matcap renderer.
pub fn register_matcap_gl_material(scene: &mut GlSceneRuntime) {
    register_gl_mesh_material_renderer(
        scene,
        matcap_material_kind(),
        Box::new(MatcapGlMeshMaterialRenderer),
    );
}

#[cfg(test)]
mod tests {
    use super::*;
    use flighthq_materials::unlit_materials::matcap_material_kind;

    use crate::gl_mesh_material_registry::get_gl_mesh_material_renderer;
    use crate::gl_scene_runtime::create_gl_scene_runtime;

    // register_matcap_gl_material

    #[test]
    fn register_matcap_gl_material_registers_for_the_matcap_kind() {
        let mut scene = create_gl_scene_runtime();
        register_matcap_gl_material(&mut scene);
        assert!(get_gl_mesh_material_renderer(&scene, matcap_material_kind()).is_some());
    }
}

File: packages/scene-gl/src/matcapGlMeshMaterialRenderer.ts
import { unpackColorToLinear } from '@flighthq/color';
import type {
  LinearColor,
  Camera3D,
  GlMeshMaterialRenderer,
  GlRenderState,
  MatcapMaterial,
  Material,
  MeshGeometry,
  SceneLightBlock,
  SceneRenderProxy,
  GlMatcapDefineKey,
} from '@flighthq/types';
import { MatcapMaterialKind } from '@flighthq/types';

import { bindGlMatcapSurface, ensureGlMatcapProgram } from './glMatcapPrelude';
import { registerGlMeshMaterialRenderer } from './glMeshMaterialRegistry';
import { beginGlMeshDraw, drawGlMeshSubset, setGlMeshViewProjection } from './glMeshProgram';
import { getGlSceneRuntime } from './glSceneRuntime';

// The built-in Matcap forward renderer (GlMeshMaterialRenderer for MatcapMaterialKind). Lighting-
// independent material-capture shading: bind selects the matcap variant for the material's matcap
// texture / alpha mode, uploads the camera view-projection plus the camera view matrix (u_view, which
// the vertex stage uses to rotate the world-space normal into view space), and the linear tint; draw
// issues the indexed draw. Lights are ignored — the matcap texture is the prebaked lighting. See
// registerMatcapGlMaterial to install it.
export const matcapGlMeshMaterialRenderer: GlMeshMaterialRenderer = {
  bind(
    state: GlRenderState,
    material: Readonly<Material> | null,
    _lights: Readonly<SceneLightBlock>,
    camera: Readonly<Camera3D>,
  ): void {
    const gl = state.gl;
    const matcap = material as Readonly<MatcapMaterial> | null;
    const program = ensureGlMatcapProgram(state, defineKeyForMaterial(matcap));
    beginGlMeshDraw(state, program, matcap !== null && matcap.doubleSided);
    setGlMeshViewProjection(gl, program.locViewProjection, camera);
    // u_view rotates the world-space normal into view space for the matcap lookup.
    gl.uniformMatrix4fv(program.locView, false, camera.view.m);

    if (matcap === null) {
      bindGlMatcapSurface(state, program, WHITE, null, 0.5);
      return;
    }
    unpackColorToLinear(scratchRgba, matcap.tint);
    bindGlMatcapSurface(state, program, scratchRgba, matcap.matcap, matcap.alphaCutoff);
  },

  draw(state: GlRenderState, proxy: Readonly<SceneRenderProxy>, geometry: Readonly<MeshGeometry>): void {
    const program = getGlSceneRuntime(state).activeMeshProgram;
    if (program === null) return;
    drawGlMeshSubset(state, program, proxy, geometry);
  },
};

// Registers the built-in Matcap renderer for MatcapMaterialKind on this state. Opt-in (no top-level
// side effect); call once per GlRenderState before drawScene so meshes with MatcapMaterials draw.
export function registerMatcapGlMaterial(state: GlRenderState): void {
  registerGlMeshMaterialRenderer(state, MatcapMaterialKind, matcapGlMeshMaterialRenderer);
}

function defineKeyForMaterial(material: Readonly<MatcapMaterial> | null): GlMatcapDefineKey {
  return {
    alphaMaskEnabled: material !== null && material.alphaMode === 'mask',
    hasMatcap: material !== null && material.matcap !== null && material.matcap.image !== null,
  };
}

const scratchRgba: LinearColor = [0, 0, 0, 0];
const WHITE: LinearColor = [1, 1, 1, 1];
